custom/nic_compliance/utils.py

```diff
@@ -0,0 +1,117 @@
+
+
+def extract_password(password):
+    # Passwords set with expected salts length and padding would respect this regex
+    reg_exp = r"^sha256\$([a-z0-9A-Z]{6})([a-zA-Z0-9=]*)([a-z0-9A-Z]{6})=$"
+    match_result = re.match(reg_exp, password)
+    # strip out outer level padding of salts/keys and ensure three matches
+    if match_result and len(match_result.groups()) == 3:
+        match_groups = match_result.groups()
+        hash_left, stripped_password, hash_right = match_groups
+        # decode the stripped password to get internal block
+        # decoded(salt1 + encoded_password + salt2)
+        try:
+            decoded_password = base64.b64decode(stripped_password)
+        except TypeError:
+            return ''
+        match_result_2 = re.match(reg_exp, decoded_password)
+        # strip out hashes from the internal block and ensure 3 matches
+        if match_result_2 and len(match_result_2.groups()) == 3:
+            match_groups = match_result_2.groups()
+            # ensure the same hashes were used in the internal block as the outer
+            if match_groups[0] == hash_left and match_groups[2] == hash_right:
+                # decode to get the real password
+                password_hash = match_groups[1]
+                # return password decoded for UTF-8 support
+                try:
+                    return base64.b64decode(password_hash).decode('utf-8')
+                except TypeError:
+                    return ''
+            else:
+                # this sounds like someone tried to hash something but failed so ignore the password submitted
+                # completely
+                return ''
+        else:
+            # this sounds like someone tried to hash something but failed so ignore the password submitted
+            # completely
+            return ''
+    else:
+        # return the password received AS-IS
+        return password
+
+
+def hash_password(password):
+    return PASSWORD_HASHER.encode(password, PASSWORD_HASHER.salt())
+
+
+def verify_password(password, password_salt):
+    return PASSWORD_HASHER.verify(password, password_salt)
+
+
+def login_attempts_redis_key_for_user(username):
+    return REDIS_LOGIN_ATTEMPTS_LIST_PREFIX + username
+
+
+def get_login_attempts(username):
+    client = get_redis_client()
+    return client.get(login_attempts_redis_key_for_user(username), [])
+
+
+def get_decoded_password(password_hash, username=None):
+    def replay_attack():
+        # Replay attack where the same hash used from previous login attempt
+        login_attempts = get_login_attempts(username)
+        for login_attempt in login_attempts:
+            if verify_password(password_hash, login_attempt):
+                return True
+
+    def record_login_attempt():
+        client = get_redis_client()
+        login_attempts = client.get(login_attempts_redis_key_for_user(username), [])
+        key_name = login_attempts_redis_key_for_user(username)
```

Comment:
could move this up one line and reuse it rather than calling `login_attempts_redis_key_for_user` twice